core/src/test/scala/org/apache/spark/HeartbeatReceiverSuite.scala

```diff
@@ -207,6 +210,54 @@ class HeartbeatReceiverSuite
     assert(fakeClusterManager.getExecutorIdsToKill === Set(executorId1, executorId2))
   }
 
+  test("expired executor should not be offered again (SPARK-24387)") {
+    scheduler = spy(new TaskSchedulerImpl(sc))
```

Comment:
Just little question want to make sure, here use spy because we need to call `runningTasksByExecutors ` later?

Reply:
Because I need a functioning scheduler rather than just a mock. And need to stub the executorHeartbeatReceived method. So I use a spy here.